datafusion/expr/src/logical_plan/builder.rs

```diff
@@ -403,6 +403,28 @@ impl LogicalPlanBuilder {
                 Ok(())
             })?;
 
+        // if current plan is distinct or current plan is repartition and its child plan is distinct,
+        // then this plan is a select distinct plan
+        let is_select_distinct = match self.plan {
+            LogicalPlan::Distinct(_) => true,
+            LogicalPlan::Repartition(Repartition { ref input, .. }) => {
+                matches!(input.as_ref(), &LogicalPlan::Distinct(_))
+            }
+            _ => false,
```

Comment:
is there no other pattern to fit this case?
just below cases?
```
   sort column1
      repartition
         distinct
```

```
   sort column1
     distinct 
```

cc @alamb

Reply:
So in my opinion this PR is better than the code on master, though I think it could be improved with the suggestions from @liukun4515 and @jackwener 

@xiaoyong-z  please let us know what you want to do -- we can merge and improve in a follow on or we can implement the suggestions in this PR

Reply:
I have added the sqllogictest and integration_test in this pr.

As you said, there may be some problems if you use DataFrame and LogicalPlanBuilder to build the plan.
>  if it is built directly (via a DataFrame or someone using LogicalPlanBuilder directly) it is probably incorrect to throw an error here (as they may have meant to to run the distinct before the sort for some reason

How about merge this pr currently, and i will improve it in a follow on pr.